subsys/logging/log_backend_fs.c

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <logging/log_backend.h>
+#include "log_backend_std.h"
+#include <assert.h>
+#include <fs/fs.h>
+#if CONFIG_LOG_BACKEND_FS_AUTOSTART
+#include <fs/littlefs.h>
+#endif
+
+#define MAX_PATH_LEN 256
+#define MAX_FLASH_WRITE_SIZE 256
+#define LOG_PREFIX_LEN (sizeof(CONFIG_LOG_BACKEND_FS_FILE_PREFIX) - 1)
+
+enum backend_fs_state {
+	BACKEND_FS_OK = 0,
+	BACKEND_FS_CORUPTED,
+	BACKEND_FS_NOT_INITIALIZED
+};
+
+#if CONFIG_LOG_BACKEND_FS_AUTOSTART
+FS_LITTLEFS_DECLARE_DEFAULT_CONFIG(storage);
+static struct fs_mount_t lfs_storage_mnt = {
+	.type = FS_LITTLEFS,
+	.fs_data = &storage,
+	.storage_dev = (void *)FLASH_AREA_ID(storage),
+	.mnt_point = CONFIG_LOG_BACKEND_FS_MOUNT_POINT,
+};
+#endif
+
+static char fname[MAX_PATH_LEN];
+static struct fs_file_t file;
+static enum backend_fs_state backend_state = BACKEND_FS_NOT_INITIALIZED;
+static int file_ctr;
+
+static int allocate_new_file(struct fs_file_t *file);
+
+#if CONFIG_LOG_BACKEND_FS_OVERWRITE
+static int del_oldest_log(void);
+#endif
+
+static int write_log_to_file(uint8_t *data, size_t length, void *ctx)
+{
+	if (backend_state == BACKEND_FS_NOT_INITIALIZED) {
+		int rc = allocate_new_file(&file);
+
+		if (rc < 0) {
+			backend_state = BACKEND_FS_CORUPTED;
+		}
+	}
+
+	if (backend_state == BACKEND_FS_OK) {
+		int rc;
+		struct fs_file_t *f = &file;
+
+		/* Check if new data overwrites max file size.
+		 * If so, create new log file.
+		 */
+		int size = fs_tell(f);
+
+		if (size < 0) {
+			backend_state = BACKEND_FS_CORUPTED;
+
+			return length;
+		} else if ((size + length) > CONFIG_LOG_BACKEND_FS_FILE_SIZE) {
+			rc = allocate_new_file(f);
+
+			if (rc < 0) {
+				backend_state = BACKEND_FS_CORUPTED;
+
+				return length;
+			}
+		}
+
+		rc = fs_write(f, data, length);
+		if (rc == -ENOSPC) {
+			if (IS_ENABLED(CONFIG_LOG_BACKEND_FS_OVERWRITE)) {
+				del_oldest_log();
+
+				return 0;
+			}
+			/* If overwrite is disabled, full memory
+			* is equivalent of corrupted backend.
+			*/
+			backend_state = BACKEND_FS_CORUPTED;
+
+			return length;
+		} else if (rc < 0) {
+			return 0;
+		}
+
+		fs_sync(f);
+	}
+
+	return length;
+}
+
+static uint8_t __aligned(4) buf[MAX_FLASH_WRITE_SIZE];
+LOG_OUTPUT_DEFINE(log_output, write_log_to_file, buf, MAX_FLASH_WRITE_SIZE);
+
+static void put(const struct log_backend *const backend,
+		struct log_msg *msg)
+{
+	log_backend_std_put(&log_output, 0, msg);
+}
+
+static void log_backend_fs_init(void)
+{
+	int rc;
+#if CONFIG_LOG_BACKEND_FS_AUTOSTART
+	/* If autostart, mount file system automatically. */
+	struct fs_mount_t *mp = &lfs_storage_mnt;
```

Comment:
ideally you could use:
```
if(IS_ENABLED(CONFIG_LOG_BACKEND_FS_AUTOSTART) {
...
}
```
and avoid ifdefs here and in struct declaration.

Reply:
Is it preferred to use `if(IS_ENABLED(...))` than `#if ...`? In the second option `static struct fs_mount lfs_storage_mnt` is not compiled (less size).

Reply:
in both cases it won't be placed in rom/ram. Advantage of `if (IS_ENABLED()` is that things get compiled. If variable is static and not used in the file it will be removed. After preprocessing compiler gets:
```
static int foo;

if (0) {
  foo = ...
}

if (0) {
  if (foo) ...
}
```
compiler figures out that foo is never used and removes it.